<?xml version="1.0"?>
<launch>
  <!-- Gazebo & GUI Configuration -->
  <arg name="gazebo"      default="true"  doc="Should the gazebo simulation be launched? Use false in case if you want to include this file and launch gazebo yourself" />
  <arg name="headless"    default="false" doc="Should the gazebo GUI be launched?" />
  <arg name="paused"      default="false" doc="Should the simulation directly be stopped at 0s?" />
  <arg name="world"       default="worlds/empty.world" doc="Filename to a SDF World for gazebo to use" />
  <arg name="rviz"        default="false" doc="Should RVIz be launched?" />

  <!-- Robot Customization -->
  <arg name="arm_id"      default="za"    doc="Name of the robot to spawn" />
  <arg name="controller"  default=" "     doc="Which example controller should be started? (One of {effort_joint_trajectory_controller})" />
  <arg name="initial_joint_positions"
       doc="Initial joint configuration of the za6. Specify as a list of name/value pairs in form of '-J [name-of-joint] [value-in-rad]'. Default is a 90 degree bend in the elbow"
       default="-J $(arg arm_id)_joint_1 0
                -J $(arg arm_id)_joint_2 0.53
                -J $(arg arm_id)_joint_3 0.47
                -J $(arg arm_id)_joint_4 0
                -J $(arg arm_id)_joint_5 -1
                -J $(arg arm_id)_joint_6 0"
       />

  <include file="$(find gazebo_ros)/launch/empty_world.launch" if="$(arg gazebo)">
    <arg name="world_name" value="$(arg world)"/>
    <!-- Always start in paused mode, and only unpause when spawning the model -->
    <arg name="paused" value="true"/>
    <arg name="gui" value="$(eval not arg('headless'))"/>
    <arg name="use_sim_time" value="true"/>
  </include>
  
  <param name="robot_description" command="xacro $(find za_description)/urdf/za.xacro prefix:=$(arg arm_id)_" />

  <rosparam file='$(find za_gazebo)/config/za_hw_sim.yaml' subst_value="true" />
  <rosparam file='$(find za_gazebo)/config/sim_controllers.yaml' subst_value="true" />

  <arg name="unpause" value="$(eval '' if arg('paused') else '-unpause')" />
  <node name="$(arg arm_id)_model_spawner"
        pkg="gazebo_ros"
        type="spawn_model"
        args="-param robot_description -urdf -z 0.01 -model $(arg arm_id) $(arg unpause)
        $(arg initial_joint_positions)
        "/>
  />

  <!-- spawns the controller after the robot was put into its initial joint pose -->
  <node pkg="controller_manager"
        type="spawner"
        name="$(arg arm_id)_controller_spawner"
        respawn="false" output="screen"
        args="za_state_controller $(arg controller)"
  />

  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
  <node name="joint_state_publisher" type="joint_state_publisher" pkg="joint_state_publisher">
    <rosparam param="source_list">[za_state_controller/joint_states]</rosparam>
    <param name="rate" value="30"/>
  </node>

</launch>